CGEN error with sum type and optional string

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 bd8d8ac, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 45fd7eb.bd8d8ac
|:-------------------|:-------------------
|OS                  |windows, Microsoft Windows 11 Pro 26100 64-bit
|Processor           |24 cpus, 64bit, little endian, AMD Ryzen 9 5900X 12-Core Processor  
|Memory              |84.74GB/127.92GB
|                    |
|V executable        |S:\repo\vlang\v.exe
|V last modified time|2025-02-04 15:08:30
|                    |
|V home dir          |OK, value: S:\repo\vlang
|VMODULES            |OK, value: C:\Users\john3\.vmodules
|VTMP                |OK, value: S:\temp\v_0
|Current working dir |OK, value: S:\vProjects\testit
|                    |
|Git version         |git version 2.45.1.windows.1
|V git status        |weekly.2025.05-73-gbd8d8ac4
|.git/config present |true
|                    |
|cc version          |N/A
|gcc version         |gcc (MinGW-W64 x86_64-ucrt-posix-seh, built by Brecht Sanders, r2) 14.2.0
|clang version       |(built by Brecht Sanders, r2) clang version 19.1.1
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |thirdparty-windows-amd64 b425ac82
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg -cg testit.v && S:\vProjects\testit\testit.exe`
```v
type Any = ?int | ?string

struct TestConfig {
	timeout ?int
}

struct Test {
	client_config TestConfig
}

fn main() {
	mut r := Test{
		client_config: TestConfig{
			timeout: 0
		}
	}
	mut pool_manager_args := map[string]?Any{}
	pool_manager_args['timeout'] = r.client_config.timeout
	pool_manager_args['New String Arg'] = ?Any('a string')
}
```

**What did you see?**
```
S:/temp/v_0/testit.01JK8S7T2W1B2JACG4CSG43M3M.tmp.c:530: warning: WINVER redefined

S:/temp/v_0/testit.01JK8S7T2W1B2JACG4CSG43M3M.tmp.c:4048: warning: implicit declaration of function 'tcc_backtrace'

S:/temp/v_0/testit.01JK8S7T2W1B2JACG4CSG43M3M.tmp.c:6900: error: cannot convert 'struct string' to 'struct _option_int *'
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

clean compile



> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -115,9 +115,17 @@ fn (mut g Gen) expr_opt_with_cast(expr ast.Expr, expr_typ ast.Type, ret_typ ast.
 				g.write('_option_ok(&(${styp}[]) {')
 			}
 			if expr is ast.CastExpr && expr_typ.has_flag(.option) {
-				g.write('*((${g.base_type(expr_typ)}*)')
-				g.expr(expr)
-				g.write('.data)')
+				ret_sym := g.table.sym(ret_typ)
+				if ret_sym.kind == .sum_type {
+					exp_sym := g.table.sym(expr_typ)
+					fname := g.get_sumtype_casting_fn(expr_typ, ret_typ)
+					g.call_cfn_for_casting_expr(fname, expr, ret_typ.is_ptr(), ret_sym.cname,
+						expr_typ.is_ptr(), exp_sym.kind == .function, g.styp(expr_typ))
+				} else {
+					g.write('*((${g.base_type(expr_typ)}*)')
+					g.expr(expr)
+					g.write('.data)')
+				}
 			} else {
 				old_inside_opt_or_res := g.inside_opt_or_res
 				g.inside_opt_or_res = false

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2585,7 +2585,7 @@ struct SumtypeCastingFn {
 
 fn (mut g Gen) get_sumtype_casting_fn(got_ ast.Type, exp_ ast.Type) string {
 	mut got, exp := got_.idx_type(), exp_.idx_type()
-	i := int(got) | int(u32(exp) << 17) | int(u32(exp_.has_flag(.option)) << 16)
+	i := int(got) | int(u32(exp) << 18) | int(u32(exp_.has_flag(.option)) << 17) | int(u32(got_.has_flag(.option)) << 16)
 	exp_sym := g.table.sym(exp)
 	mut got_sym := g.table.sym(got)
 	cname := if exp == ast.int_type_idx {

--- a/vlib/v/tests/options/option_sumtype_option_variant_test.v
+++ b/vlib/v/tests/options/option_sumtype_option_variant_test.v
@@ -1,0 +1,12 @@
+type Any = ?int | string | ?string
+
+fn test_main() {
+	b := ?Any(?string('bar'))
+	assert b?.str() == "Any(Option('bar'))"
+
+	c := ?Any(string('baz'))
+	assert c?.str() == "Any('baz')"
+
+	d := ?Any('baz')
+	assert d?.str() == "Any('baz')"
+}

